File: fb/257.binary-tree-paths.cpp
// Time O(n)
// Space O(n)
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        if(!root) return ans;
        binaryTreePaths(root, ans, "");
        return ans;
    }

private:
    void binaryTreePaths(TreeNode* root, vector<string>& paths, string path) {
        if(!root->left && !root->right) {
            path += to_string(root->val);
            paths.push_back(path);
            return;
        }
        path += to_string(root->val) + "->";
        if(root->left)
            binaryTreePaths(root->left, paths, path);
        if(root->right)
            binaryTreePaths(root->right, paths, path);
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

File: fb/257.binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        self.ans = []
        
        def dfs(node, path):
            if not node:
                return
            path += str(node.val)
            
            dfs(node.left, path + "->")
            dfs(node.right, path + "->")
            
            if not node.left and not node.right:
                self.ans.append(path)
                
        dfs(root, "")
        return self.ans
